rfd/0147-repo-mirror.md

```diff
@@ -0,0 +1,73 @@
+---
+authors: Mike Jensen (mike.jensen@goteleport.com)
+state: draft
+---
+
+# RFD 144 - Automatic Repo Mirroring
+
+## What
+
+This RFD outlines a method for continuously and automatically syncing a GitHub repository. This can be used to create private copies of a repository.
+
+## Why
+
+We currently face a few frictions which could be addressed through a more robust repository sync mechanism:
+1. Our internal `teleport-private` mirror of `teleport`, used for security development, currently requires manual updates to sync.
+2. Certain tools, like Dependabot, don't support scanning non-default branches, necessitating manual efforts to manage our release branches.
+3. Due to the committing of CI and scanning configurations in the repository, we receive duplicate alerts when using a repository for testing or as an internal mirror/fork.
+
+While addressing these immediate needs, the aim of this RFD is to establish a generic pattern for repository mirroring. This will provide flexibility for future requirements when syncing any upstream repository to a fork or mirror that we maintain.
+
+## Details
+
+The core requirements of this solution are:
+* Automatic and regular updates to a repository to ensure is's state matches the upstream.
+* The ability to make additional changes to the mirror repository, separate from the upstream repository (e.g., disabling actions we don't want to run on the mirror, or enabling additional scanning/actions).
+
+### Branch Structure
+
+* `master` (configurable) - This will serve as the repository's default branch. It will contain both the upstream history and any commits necessary for the synchronization process or other custom changes made to the mirror repository.
+* `sync/upstream-[UPSTREAM_BRANCH_NAME]` - This branch will be an exact copy of the upstream branch. Should there be a need to create a PR in this repository, the history should be based on this branch, not on `master`.
+* `sync/rebase` - This is the branch where custom changes will be committed. It will be rebased onto the upstream changes before being merged into master.
```

Comment:
On a second pass I'm not sure what the purpose of this is. If `master` (or `teleport`) is the sum of the upstream plus patches, shouldn't we just land any changes straight there?

Reply:
Any changes landed there will get overwritten on the next sync run.  In general I expect the flow to go like this:
* PR is opened on private repo
* Private release is prepared once PR is approved
* After open disclosure of issue OSS PR is opened
* Once fix is incorporated in OSS, it will be automatically synced to the mirror repo

The challenge with merging into the mirror repo is that we now have two sources of truth, in a divergent code base.  So in general I expect PR's to just be closed in the mirror repo

Let me know your thoughts.

Reply:
Gotcha, I was thinking of a different kind of fork. Sounds good.

Reply:
You make a good point, this probably never makes sense for the `teleport` branches in our forks.  We could use it to keep an upstream branch in sync, but we likely want to submit PR's and do merge commits for external forks which we want to synchronized on top of our changes.